reader-privacy.md

```diff
@@ -0,0 +1,199 @@
+    participant provider
+    client->>client: calculates hash(MH)
+    client->>indexer: sends a find request for hash(MH)
+    indexer->>client: sends a list of [ProviderRecordKey], each encrypted with a key derived from MH
+    loop ProviderRecordKeys
+        client->>client: decrypts ProviderRecordKey and extracts peerID from it
+        client->>indexer: sends ProviderRecord lookup request for hash(peerID)
+        indexer->>client: sends a ProviderRecord encrypted with a key derived from peerID
+        client->>client: decrypts the ProviderRecord
+    end
+    client->>provider: reaches out to the provider for the desired content
+```
+
+
+### Security
+​
+Security model of the Reader Privacy proposal boils down to inability to *algorithmically* derive the original multihash value for a 
+`hash(multihash)` that is used for IPNI lookups. Right now advertisments are not encrypted, but authenticated and contain plain multihash values in them. 
+That is going to change once *Writer Privacy* is implemented. Until then, an attacker could build a map of `hash(multihash) -> multihash` 
+by re-ingesting advertisements chain from each publisher in order to collect all original multihashes which can then be used to decrypt provider records and so on. 
+Doing that will require significant resources as it involves crawling the entire network. However, it will eventually be eliminated by *Writer Privacy* upgrade.
+
+Reader Privacy is a first step towards fully private content routing protocol. 
+
+Wider security implications are discussed in the IPFS Reader Privacy specification: TODO link here.
+
+#### Hashing and Encryption Function Upgrades
+
+All multihashes have a codec encoded in them. If a hashing or encryption funciton will have to rotate then different types of multihahses can coexist together
+and can be processed differently by IPNI implementations. It won't be possible to apply a fix retroactivelly to the data returned by previous lookup requests, 
+however IPNI implementations should start blocking all new ones that use a compromised scheme.
+
+Moving an IPNI implementation to a new hash / encryption function will involve reingesting all data from a scratch. Before Writer Privacy is impemented the 
+index can be migrated over to new functions by reingesting all advertisement chains. With Writer Privacy, Publishers will have to republish advertisments 
+using new algorithms. Both old and new scheme can coexist together for some time. The old one should be retired either immediately or once
+the indexes have been rebuilt and the users have been migrated over. 
+
+Exact operation procedure will be different for differnet IPNI implementations.
+
+### Trade Offs 
+
+* **Multiple lookups**. In the simplest scenario Reader Privacy protocol will require at least two roundtrips to find provider details for a given CID.
+It can be reduced down to one by caching `ProviderRecord`s locally at the client side. That would eliminate a need in a lookup
```

Comment:
Yes, exactly. This is what `ProviderRecord` is - peer id combined with multiaddresses and some other metadata (like supported protocols)

Reply:
Alright I got confused because we use the name `ProviderRecord` for the `CID -> PeerID` mapping (or `Hash(mulithash) -> PeerID`) and the name `PeerRecord` for the `PeerID -> Multiaddr` mapping.

The `ProviderRecordKey` and `ProviderRecord` are well defined in the _Specification_ section, I must have missed it while reading 👍🏻